bittide/src/Bittide/Transceiver.hs

```diff
@@ -7,453 +7,556 @@
   , HasDefinedInitialValues tx
 
   , HasSynchronousReset rx
   , HasDefinedInitialValues rx
 
-  , HasSynchronousReset freeclk
-  , HasDefinedInitialValues freeclk
-  ) =>
-  Clock refclk ->
-  Clock freeclk ->
-
-  Reset freeclk ->
+  , HasSynchronousReset free
+  , HasDefinedInitialValues free
 
-  Vec chansUsed String ->
-  Vec chansUsed String ->
+  , KnownDomain rxS
+  , KnownDomain txS
+  , KnownDomain ref
+  , KnownDomain free
+  ) =>
+  Config free ->
+  Inputs tx rx ref free rxS n ->
+  Outputs n tx rx txS free
+transceiverPrbsN opts inputs@Inputs{clock, reset, refClock} = Outputs
+  -- tx
+  { txClocks          = map (.txClock) outputs
+  , txResets          = map (.txReset) outputs
+  , txReadys          = map (.txReady) outputs
+  , txSamplings       = map (.txSampling) outputs
+
+  -- rx
+  , rxClocks = map (.rxClock) outputs
+  , rxResets = map (.rxReset) outputs
+  , rxDatas  = map (.rxData) outputs
 
-  Vec chansUsed (Signal rxS (BitVector 1)) ->
-  Vec chansUsed (Signal rxS (BitVector 1)) ->
+  -- transceiver
+  , txPs = pack <$> bundle (map (.txP) outputs)
+  , txNs = pack <$> bundle (map (.txN) outputs)
 
-  TransceiverOutputs chansUsed tx rx txS freeclk
-transceiverPrbsN refclk freeclk rst chanNms clkPaths rxns rxps = TransceiverOutputs
-  { txClocks = map (.txClock) outputs
-  , rxClocks = map (.rxClock) outputs
-  , txPs     = map (.txP)     outputs
-  , txNs     = map (.txN)     outputs
-  , linkUps  = map (.linkUp)  outputs
-  , stats    = map (.stats)   outputs
+  -- free
+  , linkUps    = map (.linkUp)  outputs
+  , linkReadys = map (.linkReady)  outputs
+  , stats      = map (.stats)   outputs
   }
  where
-  outputs = zipWith4 (transceiverPrbs refclk freeclk rst) chanNms clkPaths rxns rxps
+  -- XXX: Replacing 'zipWithN' with '<$>' and '<*>' triggers a combination of:
+  --
+  --       * https://github.com/clash-lang/clash-compiler/issues/2723
+  --       * https://github.com/clash-lang/clash-compiler/issues/2722
+  --
+  outputs = zipWith8 go
+    (iterateI (+1) 0)
+    inputs.channelNames
+    inputs.clockPaths
+    (unbundle (unpack <$> inputs.rxNs))
+    (unbundle (unpack <$> inputs.rxPs))
+    inputs.txDatas
+    inputs.txReadys
+    inputs.rxReadys
+
+  go transceiverIndex channelName clockPath rxN rxP txData txReady rxReady =
+    transceiverPrbs Gth.gthCore opts Input
+      { channelName, clockPath, rxN, rxP, txData, txReady, rxReady, transceiverIndex
+      , clock, reset, refClock
+      }
 
 transceiverPrbs ::
+  forall tx rx ref free txS rxS serializedData .
   ( HasSynchronousReset tx
   , HasDefinedInitialValues tx
 
   , HasSynchronousReset rx
   , HasDefinedInitialValues rx
 
-  , HasSynchronousReset freeclk
-  , HasDefinedInitialValues freeclk
-  ) =>
+  , HasSynchronousReset free
+  , HasDefinedInitialValues free
 
-  Clock refclk ->
-  Clock freeclk ->
+  , KnownDomain rxS
+  , KnownDomain txS
+  , KnownDomain ref
+  , KnownDomain free
+  ) =>
+  GthCore tx rx ref free txS rxS serializedData ->
```

Comment:
Why do we need to provide a core for `transceiverPrbs` but not for `transceiverPrbsN`? I would expect a uniform interface here.